File: src/WirePact.PoC.Translator/Startup.cs
using DotnetKubernetesClient;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using WirePact.PoC.Translator.Interceptors;
using WirePact.PoC.Translator.Kubernetes;

namespace WirePact.PoC.Translator
{
    public class Startup
    {
        private readonly IConfiguration _configuration;

        public Startup(IConfiguration configuration)
        {
            _configuration = configuration;
        }

        public void ConfigureServices(IServiceCollection services) =>
            services
                .AddLogging(builder => builder.AddSimpleConsole().AddConfiguration(_configuration))
                .AddGrpc(
                    options =>
                    {
                        options.EnableDetailedErrors = true;
                        options.Interceptors.Add<TimeLogger>();
                    })
                .Services
                .AddSingleton<ZitadelValidator>()
                .AddSingleton<CredentialsReader>()
                .AddTransient<IKubernetesClient, KubernetesClient>();

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseRouting();

            app.UseEndpoints(
                endpoints => endpoints.MapGrpcService<Services.Translator>());
        }
    }
}

File: src/WirePact.PoC.Translator/Startup.fs
namespace WirePact.PoC.Translator

open System
open DotnetKubernetesClient
open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.Hosting
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.Logging

type Startup(config: IConfiguration) =

    member _.ConfigureServices(services: IServiceCollection) =
        services
            .AddLogging(fun builder ->
                builder
                    .AddSimpleConsole()
                    .AddConfiguration(config)
                |> ignore)
            .AddSingleton<ZitadelValidator>()
            .AddSingleton<CredentialReader>()
            .AddGrpc(fun options ->
                options.EnableDetailedErrors <- true
                options.Interceptors.Add<TimeLogger>())
        |> ignore

#if DEBUG
        services.AddTransient<IKubernetesClient>(fun _ -> Mocks.kubernetesClient)
        |> ignore

        Environment.SetEnvironmentVariable("CREDENTIALS_SECRET_NAME", "secret")
#else
        services.AddTransient<IKubernetesClient, KubernetesClient>()
        |> ignore
#endif

        ()

    member _.Configure(app: IApplicationBuilder, env: IWebHostEnvironment) =
        if env.IsDevelopment() then
            app.UseDeveloperExceptionPage() |> ignore

        app
            .UseRouting()
            .UseEndpoints(fun endpoints -> endpoints.MapGrpcService<Translator>() |> ignore)
        |> ignore
